test/jdk/java/lang/ProcessBuilder/ReaderWriterTest.java

```diff
@@ -0,0 +1,225 @@
+ * @build jdk.test.lib.process.ProcessTools jdk.test.lib.hexdump.HexPrinter
+ * @run testng ReaderWriterTest
+ */
+
+@Test
+public class ReaderWriterTest {
+
+    static final String ASCII = "ASCII: \u0000_A-Z_a-Z_\u007C_\u007D_\u007E_\u007F_;";
+    static final String ISO_8859_1 = " Symbols: \u00AB_\u00BB_\u00fc_\u00fd_\u00fe_\u00ff;";
+    static final String FRACTIONS = " Fractions: \u00bc_\u00bd_\u00be_\u00bf;";
+
+    public static final String TESTCHARS = "OneWay: " + ASCII + ISO_8859_1 + FRACTIONS;
+    public static final String ROUND_TRIP_TESTCHARS = "RoundTrip: " + ASCII + ISO_8859_1 + FRACTIONS;
+
+    @DataProvider(name="CharsetCases")
+    static Object[][] charsetCases() {
+        return new Object[][] {
+                {"UTF-8"},
+                {"ISO8859-1"},
+                {"US-ASCII"},
+        };
+    }
+
+    /**
+     * Test the defaults case of native.encoding.  No extra command line flags or switches.
+     */
+    @Test
+    void testCaseNativeEncoding() throws IOException {
+        String nativeEncoding = System.getProperty("native.encoding");
+        Charset cs = Charset.forName(nativeEncoding);
+        System.out.println("Native.encoding Charset: " + cs);
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder("ReaderWriterTest$ChildWithCharset");
+        Process p = pb.start();
+        writeTestChars(p.outputWriter(true));
+        checkReader(p.inputReader(), cs, "Out");
+        checkReader(p.errorReader(), cs, "Err");
+        try {
+            int exitValue = p.waitFor();
+            if (exitValue != 0)
+                System.out.println("exitValue: " + exitValue);
+        } catch (InterruptedException ie) {
+            Assert.fail("waitFor interrupted");
+        }
+    }
+
+    /**
+     * Write the test characters to the child using the Process.outputWriter.
+     * @param writer the Writer
+     * @throws IOException if an I/O error occurs
+     */
+    private static void writeTestChars(Writer writer) throws IOException {
+        // Write the test data to the child
+        try (writer) {
+            writer.append(ROUND_TRIP_TESTCHARS);
+            writer.append(System.lineSeparator());
+        }
+    }
+
+    /**
+     *
+     * @param nativeEncoding
+     */
+    @Test(dataProvider = "CharsetCases", enabled = true)
+    void testCase(String nativeEncoding) throws IOException {
+        String osName = System.getProperty("os.name").toLowerCase(Locale.ROOT);
+
+        Charset cs = Charset.forName(nativeEncoding);
+        System.out.println("Charset: " + cs);
+        String cleanCSName = cleanCharsetName(cs);
+
+        ProcessBuilder pb = ProcessTools.createJavaProcessBuilder(
+                "-Dsun.stdout.encoding=" + cleanCSName,     // Encode in the child using the charset
+                "-Dsun.stderr.encoding=" + cleanCSName,
+                "ReaderWriterTest$ChildWithCharset");
+        var env = pb.environment();
+        env.put("LANG", "en_US." + cleanCSName);
```

Comment:
Does this work on Windows?

Reply:
Should be removed, the tests work because they set sun.stdout/stderr.encoding.